src/components/d3/gauge-arc.js

```diff
@@ -0,0 +1,106 @@
+import PropTypes from 'prop-types'
+import React, { useRef, useEffect, useMemo } from 'react'
+import * as d3 from 'd3'
+
+const GaugeArc = ({ data, width, height, config }) => {
+
+  const { color, backgroundColor, dataKey, transition } = config
+
+  const value = data[dataKey.x2] ? data[dataKey.x1] / data[dataKey.x2] * 100 : 0
+
+  const remainder = 100 - value
+  const colors = [color, backgroundColor]
+  const svgRef = useRef(null)
+
+  const isAscending = value > remainder
+
+  const pieSize = width > height ? height / 2 : width / 2
+
+  const lineThickness = pieSize * 0.2
+
+  const arcsGenerator = d3.pie().sortValues(function (a, b) { return isAscending ? b - a : a - b })
+
+  const arcsData = useMemo(() => arcsGenerator([value, remainder]), [data])
+
+  const arcPath = d3.arc().innerRadius(pieSize - lineThickness).outerRadius(pieSize).cornerRadius(lineThickness)
+
+  const [backgroundPath] = d3.pie()([1.0])
+
+  useEffect(() => {
+    if (svgRef.current !== null) {
+      const svg = d3.select(svgRef.current)
+
+      const gauge = svg
+        .selectAll('.arcs')
+        .data(arcsData)
+        .attr('d', arcPath)
+        .attr('fill', (_, i) => colors[i])
+
+      if (transition.animation) {
+        gauge
+          .transition()
+          .duration(transition.duration)
+          .attrTween('d', function (d) {
+            if (d.index === 1) {
+              return //deactivate tweening for the remainder value
+            }
+            var i = d3.interpolate(d.startAngle + 0.1, d.endAngle)
+            return function (t) {
+
+              d.endAngle = i(t)
+              return arcPath(d)
+            }
+          })
+      }
+    }
+  }, [svgRef.current, arcsData])
```

Comment:
I don't see any usecases which the transition prop changes after the first cycle as the animation should be triggered only once.

Reply:
My understanding is that the rules are only the recommendation made without a context of the code itself which can alter the way it functions, therefore it is only set to give us warning not error. For the same component, adding all dependencies from what eslint recommended messed up with the rendering cycle so I had to revert in overlord.

Reply:
Well, it is not what react docs actually state. This is from the official React page, at: https://reactjs.org/docs/hooks-effect.html
`Note
If you use this optimization, make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders. Learn more about how to deal with functions and what to do when the array changes too often.

If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument. This tells React that your effect doesn’t depend on any values from props or state, so it never needs to re-run. This isn’t handled as a special case — it follows directly from how the dependencies array always works.

If you pass an empty array ([]), the props and state inside the effect will always have their initial values. While passing [] as the second argument is closer to the familiar componentDidMount and componentWillUnmount mental model, there are usually better solutions to avoid re-running effects too often. Also, don’t forget that React defers running useEffect until after the browser has painted, so doing extra work is less of a problem.

We recommend using the exhaustive-deps rule as part of our eslint-plugin-react-hooks package. It warns when dependencies are specified incorrectly and suggests a fix.
`
Also, I have faced this in many code reviews in the past year and always found a better solution. I don't know if it is possible in this case. Maybe @woozyking can give a better input.

Reply:
It states that the purpose of including all values used in useEffect is to prevent itself from not re-rendering when the value updates. Now when you take a look at this deeper, many of our system cause an unnecessary rendering due to its data flow. Geocohort report tab for example, it re renders when it fetches other non-geocohort data such as conversion or overview, and react recognizes it as a mutated value when geocohort data itself hasn't changed. So in order to manually orchestrate the rendering cycle in harmony, some manual adjustment has to be imposed. Selective deps is the one way. Another way of addressing that while avoiding warning is to add conditional statement but I am not a big supporter of adding bunch of them as it defeats the purpose of hooks dependencies.

Reply:
It states that the reason that they recommend adding all values is to prevent the case where component does not update due to its stale value. However in this case, I purposefully do not want to update component because it has to animate only once in the first rendering cycle. One way to address I know would be imposing conditional statement inside of useEffect but I personally think that having too many of them are defeating the purpose of having deps list while it makes the code bulky. Therefore I am leaning towards this approach.